### PR TITLE
feat: parallel queries

### LINKED ISSUE #388: Parallel Queries
Right now, Mosaic sends one query after another to the server. This means that it always has to wait for a result to return before a next query is sent which means the server is not used effectively. As a first step, try to send a batch of queries. By batching we still maintain the order or queries which avoids potential bugs with out-of-order responses. In the next step, try building a HTTP server that processes requests in parallel and returns results out of order without leading to bugs in the UI.

### PR DESCRIPTION
**Pull request by domoritz**
Updated version of #438

Rather than sending requests in series, send them in parallel but serialize the order before returning them from the query manager. 

Fixes #388

### CODE REVIEW

**Comment by domoritz** [comment]
Some quick experiments on localhost. 

Before

node:
Screenshot 2024-08-05 at 10 42 07

python:

Screenshot 2024-08-05 at 10 46 02

After

node:
Screenshot 2024-08-05 at 10 40 35

python:

Screenshot 2024-08-05 at 10 45 31

**Comment by domoritz** [comment]
I didn't do an in-depth analysis but in some quick comparisons it looks like very results are resolved more quickly. The network manager also shows queries running in parallel.

Here is an example with 2G network simulation in Firefox

Before: 
https://github.com/user-attachments/assets/bad87032-e756-49b0-8171-5ad3df0c0aac

After:
https://github.com/user-attachments/assets/9fc46638-c536-4e56-aaec-ef6ad0a374d9

File: packages/core/test/coordinator-test.js
Comment (jheer) [inline packages/core/test/coordinator-test.js:1]: This file changes the default quote from single to double. Can we make this consistent with other files?

File: .vscode/launch.json
Comment (jheer) [inline .vscode/launch.json:1]: Don't see why this file is part of the PR.

File: eslint.config.js
Comment (jheer) [inline eslint.config.js:1]: Don't see why this file is part of the PR.

File: packages/core/src/util/query-result.js
Comment (jheer) [inline packages/core/src/util/query-result.js]: Typo here.

File: packages/core/src/util/query-result.js
Comment (jheer) [inline packages/core/src/util/query-result.js]: `QueryState.ready` might be a reasonable alternative to `prepared`?

File: packages/core/src/QueryManager.js
Comment (jheer) [inline packages/core/src/QueryManager.js]: It's a separate issue, but this makes me think we should double check caching interactions with exec queries. I don't think we want to be going to the cache for any exec queries...

**Review by jheer — CHANGES_REQUESTED** [review]
LGTM, once comments are addressed.

File: .vscode/launch.json
Comment (domoritz) [reply .vscode/launch.json:1]: Moved to [#489](https://github.com/uwdata/mosaic/pull/489)

File: eslint.config.js
Comment (domoritz) [reply eslint.config.js:1]: Pulled out into [#490](https://github.com/uwdata/mosaic/pull/490)

File: packages/core/src/util/query-result.js
Comment (domoritz) [reply packages/core/src/util/query-result.js]: Sounds good. Should the method still be called `prepare` or become `ready`?

File: packages/core/src/QueryManager.js
Comment (domoritz) [reply packages/core/src/QueryManager.js]: It looks like [cache is set to false by default](https://github.com/uwdata/mosaic/blob/76e4874f05eacb27452e85f49c38d5adddd3dfbf/packages/core/src/QueryManager.js#L40) and `exec` [makes a query without cache](https://github.com/uwdata/mosaic/blob/76e4874f05eacb27452e85f49c38d5adddd3dfbf/packages/core/src/Coordinator.js#L118) so it should always be off.

Comment (jheer) [reply packages/core/src/QueryManager.js]: Good, sounds like younger me did think it through :)

File: packages/core/src/util/query-result.js
Comment (jheer) [reply packages/core/src/util/query-result.js]: I'd rename it to `ready`

Comment (domoritz) [reply packages/core/src/util/query-result.js]: done

**Comment by domoritz** [comment]
I tested the (at least casually) the flight 200k and 10m examples on

* node sever socket and rest
* python server socket and rest
* wasm

There is a problem when I brush on one chart and move the cursor over another chart because mosaic then tries to run the query to build the index and query the index at the same time (and the latter fails since the index is not there yet).

I guess this problem (index not ready yet) can also happen if the index creation is slow and you start interacting without the index being there. I think the solution here is to always prioritize exec queries and don't send any queries until the exec is done. I'll implement that unless you have another suggestion. We may still want to prevent activation of chart while the mouse is down. At some point it might be nice to have an actual dependency model for queries but that's a lot more invasive.

**Comment by jheer** [comment]
> I guess this problem (index not ready yet) can also happen if the index creation is slow and you start interacting without the index being there. I think the solution here is to always prioritize exec queries and don't send any queries until the exec is done. I'll implement that unless you have another suggestion. We may still want to prevent activation of chart while the mouse is down. At some point it might be nice to have an actual dependency model for queries but that's a lot more invasive.

I think blocking on `exec` queries probably makes sense. There are other specs with derived data sets on load that also assume serialization -- though in some cases multiple exec queries may be issued as a single DB request, I don't think this is always the case.

I agree that a full dependency model is a bigger lift. That said, checking (and if necessary, blocking) on data cube table construction within the data cube indexer might not be so bad as a stop-gap.

Either way, we should make sure we have accompanying unit tests for these cases.

**Comment by domoritz** [comment]
Added logic to prevent multiple parallel exec requests and also added a unit test.

File: packages/core/src/QueryManager.js
Comment (jheer) [inline packages/core/src/QueryManager.js]: Shouldn't this call route through the logger?

File: packages/core/src/QueryManager.js
Comment (jheer) [inline packages/core/src/QueryManager.js]: This line looks harmless, but raises a question: do we ever see exec calls hit the cache? I don't think we should... they should be issued with `cache: false | undefined`. And if we didn't shouldn't the cache result resolve to `undefined` anyway? So I don't think we need this conditional.

File: packages/core/src/QueryManager.js
Comment (jheer) [inline packages/core/src/QueryManager.js:113]: Same comment as above.

File: packages/duckdb-server-rust/src/app.rs
Comment (jheer) [inline packages/duckdb-server-rust/src/app.rs:45]: Once #519 lands data cube indexes will no longer be temp tables. But other Mosaic exec calls may make temp tables, so I don't think we can expand the connection pool...

**Review by jheer — APPROVED** [review]
Overall looks good to me, pending my review comments. _But please do not merge until after #519 lands and I've cut a v0.11.0 release_. I plan to cut a release without parallel queries and then merge this PR for the subsequent release, in part to permit more careful testing.

In addition, we need to check all of the database connectors. I think `rest` and `wasm` should be fine. I didn't re-check the Jupyter widget but I think it will be ok for parallel use, too. However, the `socket` connector includes its own blocking, forcing sequential query execution. We need to revisit that.

File: packages/core/src/QueryManager.js
Comment (domoritz) [reply packages/core/src/QueryManager.js]: Done

File: packages/core/src/QueryManager.js
Comment (domoritz) [reply packages/core/src/QueryManager.js]: Right, we don't cache exec requests but actually could benefit from it. Below, we add the promise for pending requests to the cache so we don't submit the same request twice.

This makes me think that we maybe want to have a separate map of pending requests that we put requests in no matter whether their results should be cached or not.

Comment (jheer) [reply packages/core/src/QueryManager.js]: I don’t think it is safe to cache exec queries, as they can modify DB state. For example if I add a table, drop it, then add it again (which could maybe happen for a data cube index table), I don’t want the last exec call ignored due to caching.

Comment (domoritz) [reply packages/core/src/QueryManager.js]: Good example. A common request of creating an index would be nice to cache, though. I guess for now we can not cache and hopefully improve the parallelization when we have full query dependency tracking.

File: packages/duckdb-server-rust/src/app.rs
Comment (domoritz) [reply packages/duckdb-server-rust/src/app.rs:45]: Yes, will leave this at 1 for now. 

What are your thoughts on not using temp tables at all anymore? We could change all the examples to write tables into a separate namespace so they are still easy to clean up.

File: packages/core/src/QueryManager.js
Comment (domoritz) [reply packages/core/src/QueryManager.js:113]: We need this here so that we don't err in QueryResult.fulfill.

**Comment by domoritz** [comment]
Updated the code to address your comments. 

> However, the socket connector includes its own blocking, forcing sequential query execution. We need to revisit that.

I tested with the ws connector and it works fine. We are unnecessarily trying to parallelize queries that are then executed in series by the connector, yes. I could add a flag to not parallelize in the query manager but that introduces an extra code path so I'd be in favor of merging as is.

**Comment by jheer** [comment]
> I tested with the ws connector and it works fine. We are unnecessarily trying to parallelize queries that are then executed in series by the connector, yes. I could add a flag to not parallelize in the query manager but that introduces an extra code path so I'd be in favor of merging as is.

I agree that we don't need another flag. My point was just to note that we need to update the socket connector to actually get benefits of parallel queries there.

**Comment by domoritz** [comment]
Yes, that would be nice. I don't see an obvious way right now to do that, though. When we get a result on the socket, we just send binary so we can't easily change the order of responses. We would need to add some control signal that's sent before a result to send query results out of order. Should we do this in this pull request or a follow up (I can file an issue for it)?

**Comment by domoritz** [comment]
I filed https://github.com/uwdata/mosaic/issues/547 as a follow up.

**Pull request merged**
